### DISCUSSION TITLE
Lambda type annotation idea

### DISCUSSION THREAD

**Discussion by Moosems** [modular/modular#550] [Ideas]
In the Mojo docs, it states that "Mojo does not yet support defining anonymous functions with the lambda keyword. This is just because we couldn’t decide where to put the type annotations". I have a few ideas for the possible type annotation structure. I know that Mojo will likely take the `scala3` approach and that this isn't on the short term to do but I thought I'd at least put my had in the ring.

Here are a few:
```mojo
struct x[T: Int]: ...
    fn add_nums[func: fn(T, T) -> T](): ...
...

# Because the colon has been used for normal type annotation we need to use special character(s) reserved for lambda such as ~ or -=-.
x.add_nums[lambda x: Int, y: Int -> Int ~ x+y]()
x.add_nums[lambda x: Int, y: Int -> Int -=- x+y]()
```

Of course, that seems like a very cheeky and not great way of getting this done. That is why this could also work:
```mojo
struct x[T: String]: ...
    # Uses an inner set of parentheses to show that this is a singular expression and we define the types here.
    fn make_string_num[(T, T2: String, func: fn(T) -> T2)]()
...

# Since the types are already defined in the compile time parameters we simply use the corresponding values and placement.
x.make_num_string[lambda num_string: atol(num)]()
```

I would like to make note that error handling could definitely be a  problem in the future when using methods like `atol()` as such methods *can* fail and that this should probably have a fix as well.

This leads to one possible solution:
```mojo
struct x[T: String]: ...
    # Uses an inner set of parentheses to show that this is a singular expression and we define the types here.
    fn make_string_num[(T, T2: String, func: fn(T) -> T2)]()
...

# Here we define a default value and use a keyword, say, `fixable`, as an example,
# which lets the compiler know that if the method fails to use the value after the equal sign.
# To fix possible compiler issues due to using the equal sign we also wrap this expression in parentheses 
# which would also match the method definition nicely.
let default_value: Int = 0
x.make_num_string[(lambda num_string: fixable atol(num) = default_value)]()
```

Defining a variable like this might not be the nicest looking thing though so there is also this alternative:
```mojo
struct x[T: String]: ...
    # Uses an inner set of parentheses to show that this is a singular expression and we define the types here.
    fn make_string_num[(T, T2: String, func: fn(T) -> T2)]()
...

# Similar to the previous example we might introduce a new syntax like `single` that tells
# the compiler that the value is only used once. Since the compiler also knows that the
# return type must be an Int there should be no issue using the walrus operator to set the value
# since the walrus operator would also require this operation to be in parentheses.
x.make_num_string[(lambda num_string: fixable atol(num) = (single default_value := 0)]()
```

This might make the final syntax look like this:
```mojo
struct x[T: String]: ...
    # Uses an inner set of parentheses to show that this is a singular expression and we define the types here.
    # This also makes a nicer look if using multiple parameters/arguments since the expression is fully encapsulated.
    fn make_string_num[(T, T2: String, func: fn(T) -> T2)]()

    # The lambda also could be defined in the runtime arguments instead of compile time parameters
    fn add_nums((T, T2: Int, func: fn(T2, T2) -> T))
...

x.make_num_string[(lambda num_string: fixable atol(num) = (single default_value := 0)]()
x.make_num_string[(lambda num_string: atol(num)]() # Compiler error due to possible failure
x.add_nums((lambda x, y: String(x+y))

```

There are of course many routes that could be taken but I believe that these are the most flexible as it would allow multiple different types, is fairly simple/readable, and would make the compilers life much easier. I understand if this proposal is immediately rejected and no hurt feelings or resentment will take place. I love Mojo and wish only the best for the team. If you (the team) or anyone else would like to propose a similar with modifications to the syntax, please let me know as it can be changed and I recognize that there may be better ways but I couldn't think of many better ideas myself. Again, no hurt feelings if this isn't what you're looking for. Keep up the good work.

**Comment by Moosems** [comment]
I would also note that the `scala3` lambda function as shown [here](https://docs.scala-lang.org/scala3/book/fun-anonymous-functions.html) is also quite similar to my proposal though I had only decided to look at their implementation after writing the post so I could give clean thoughts. I do believe still that the `scala3` approach might have errors with failable methods like `atol()` and since try/except can't be used this might still actually work better but be modified to be closer to the `scala3` implementation.

**Comment by Sarctiann** [comment]
First of all, I think it's important to note that the use of lambdas in python is more geared towards cases where the type is already implicitly defined. So there's no reason to type them...

outside of that if we still intend to have something like this:
`add = lambda x, y: x + y`
I think this is enough:
`let add: fn(Int, Int) -> Int = lambda x, y: x + y`

**Reply by Moosems** [reply]
> First of all, I think it's important to note that the use of lambdas in python is more geared towards cases where the type is already implicitly defined. So there's no reason to type them...

What do you think the devs meant by "we couldn’t decide where to put the type annotations"?

**Reply by czheo** [reply]
upvoted. I also think lambda is one of the least appreciated features in Python.

**Reply by Moosems** [reply]
> let add: fn(Int, Int) -> Int = lambda x, y: x + y

Can you show this as an argument?

**Reply by Sarctiann** [reply]
I think it's overthinking to try to put annotations inside an anonymous function...
as I put in my example, it would be enough to annotate the variable (when you want to assign a lambda to a variable).
Since when the lambda is used within a method, for example:
`[1,2,3,4,5,6,7,8].sort(key=lambda n: n % 2 == 0)`
the type is already implied.

Static typing has to be a tool for the developer and for the compiler. It doesn't make sense to put types where they are not required.

**Reply by Moosems** [reply]
I honestly would prefer annotations in my code either way.

**Reply by Sarctiann** [reply]
Again referring to my previous comment, you could do this:
`[1,2,3,4,5,6,7,8].sort(key: fn(Int, Int) -> Int = lambda n: n % 2 == 0)`

**Reply by Moosems** [reply]
From the docs:
```
Small independent features
There are a number of features that are missing that are important to round out the language fully, but which don’t depend strongly on other features. These include things like:

Keyword arguments in functions: foo(x=42)
```

**Reply by Sarctiann** [reply]
I'm sorry, I don't see the problem or the relationship between one thing and the other...
it's only a matter of time before they implement these things

**Comment by ksandvik** [comment]
Maybe I'm in the minority but I like the Python lamda notation as it spells things out rather than over-using the ASCII-7 character set.
